### PR TITLE
updates to P3, including a new 3-moment option

### PR DESCRIPTION
**Pull request by hmorrison100**
TYPE: new feature

KEYWORDS: P3 microphysics

SOURCE: Hugh Morrison (NCAR)

DESCRIPTION OF CHANGES:

1. Main update is to add 3-moment option for P3, which we have chosen as MP option 53. This uses the same P3 module so that all 4 versions (MP options 50, 51, 52, 53) use the same code base. Also added new lookup tables.

2. Besides changes for the 3-moment option, a small bug fix was made to lookup table and the lookup tables were generalized to allow a larger range of mean ice particle sizes. This changes answers but has little impact compared to P3 in the previous WRF release. 

3. We also did general code cleanup and improved code comments.

LIST OF MODIFIED FILES: 
module_mp_p3.F
module_microphysics_driver.F
module_physics_init.F
Registry_EM.COMMON

Also added new lookup tables in: 
run/p3_lookupTable_2.dat-4.1
run/p3_lookupTable_1.dat-2momI_v5.1.6_oldDimax
run/p3_lookupTable_1.dat-3momI_v5.1.6.gz

Deleted the old lookup tables:
p3_lookup_table_1.dat-v4.1
p3_lookup_table_2.dat-v4.1

Modified files to handle moving the p3 lookup tables around:
Makefile
clean
test/em_b_wave/run_me_first.csh
test/em_convrad/run_me_first.csh
test/em_quarter_ss/run_me_first.csh
test/em_scm_xy/run_me_first.csh
test/em_seabreeze2d_x/run_me_first.csh
test/em_tropical_cyclone/run_me_first.csh

[Editor's note: We likely are removing all of these `run_me_first.csh` scripts in a later PR.]

TESTS CONDUCTED: 
1. Extensive testing was done for all MP options for P3 (50, 51, 52, 53)
2. Jenkins test is all pass.

RELEASE NOTE: An updated version of P3 now includes a 3-moment option. Milbrandt et al. (2021) [Milbrandt, J. A., H. Morrison, D. T. Dawson II, and M. Paukert, 2021: A triple-moment representation of ice in the Predicted Particle Properties (P3) microphysics scheme, J. Atmos. Sci., 78(2), 439-458, https://doi.org/10.1175/JAS-D-20-0084.1]

### CODE REVIEW

File: Registry/Registry.EM_COMMON [hunk 1]
```diff
@@ -2823,6 +2825,7 @@ package   thompsonaero    mp_physics==28               -             moist:qv,qc
 package   p3_1category    mp_physics==50               -             moist:qv,qc,qr,qi;scalar:qni,qnr,qir,qib;state:re_cloud,re_ice,vmi3d,rhopo3d,di3d,refl_10cm,th_old,qv_old
 package   p3_1category_nc mp_physics==51               -             moist:qv,qc,qr,qi;scalar:qnc,qni,qnr,qir,qib;state:re_cloud,re_ice,vmi3d,rhopo3d,di3d,refl_10cm,th_old,qv_old
 package   p3_2category    mp_physics==52               -             moist:qv,qc,qr,qi,qi2;scalar:qnc,qni,qnr,qir,qib,qni2,qir2,qib2;state:re_cloud,re_ice,vmi3d,rhopo3d,di3d,vmi3d_2,rhopo3d_2,di3d_2,refl_10cm,th_old,qv_old
+package   p3_1cat_3mom    mp_physics==53               -             moist:qv,qc,qr,qi;scalar:qnc,qni,qnr,qir,qib,qzi;state:re_cloud,re_ice,vmi3d,rhopo3d,di3d,refl_10cm,th_old,qv_old
```
Comment (davegill) [inline Registry/Registry.EM_COMMON:2828]: Hugh,
Look in Registry.NMM
The same place where you find p3_2category package, add in p3_1cat_3mom package.
That might be the NMM build failure.

Comment (davegill) [reply Registry/Registry.EM_COMMON:2828]: Just add in Qv. We'll put in a check_a_mundo switch to say not to use this option in NMM

**Comment by davegill** [comment]
@hmorrison100 
The only failed job uses mp_physics=52.
```
 &physics
 mp_physics                          = 52,    52,    52,
```

Is this related to the new look-up table? Such as - it is not uncompressed in the regression test?

File: Registry/Registry.EM_COMMON (hunk 1 above)
Comment (hmorrison100) [reply Registry/Registry.EM_COMMON:2828]: ok done. I'm not actually sure even how to do the build in NMM to test this, but I added the qv-only dummy option for 3-moment P3 following the existing code as you suggested.

**Comment by hmorrison100** [comment]
I fixed the problem w/ option 52. The module_mp_p3.F code was pointing to the wrong lookup table (mix up with the lookup table complications w/ zipping, etc.). It works now in my tests.

**Comment by davegill** [comment]
@hmorrison100 
Hugh,
The mechanics of the regression test are now OK with your code.
```
Please find result of the WRF regression test cases in the attachment. This build is for Commit ID: 2337703f45d882ec53d71eda3fec914a0f6fcaef, requested by: hmorrison100 for PR: https://github.com/wrf-model/WRF/pull/1356. For any query please send e-mail to David Gill.

    Test Type              | Expected  | Received |  Failed
    = = = = = = = = = = = = = = = = = = = = = = = =  = = = =
    Number of Tests        : 19           18
    Number of Builds       : 48           46
    Number of Simulations  : 166           164        0
    Number of Comparisons  : 105           104        0

    Failed Simulations are: 
    None
    Which comparisons are not bit-for-bit: 
    None
```

**Comment by davegill** [comment]
@weiwangncar @dudhia 
Folks,
Hugh's 3-moment PR is ready for review. The jenkins tests are working fine.

File: Makefile [hunk 2]
```diff
@@ -121,6 +121,11 @@ wrf : framework_only
 	if [ $(ESMF_COUPLING) -eq 1 ] ; then \
 	  ( cd main ; $(MAKE) RLFLAGS="$(RLFLAGS)" MODULE_DIRS="$(ALL_MODULES)" SOLVER=em em_wrf_SST_ESMF ) ; \
 	fi
+	if [ ! -f run/p3_lookupTable_1.dat-3momI_v5.1.6 ] ; then \
+	  # ( cd run ; gunzip --keep p3_lookupTable_1.dat-3momI_v5.1.6.gz ) ; \
```
Comment (davegill) [inline Makefile:125]: delete this commented out line

**Comment by davegill** [comment]
@hmorrison100 @dudhia @weiwangncar 
This seems ready for review.

**Comment by davegill** [comment]
@hmorrison100 @weiwangncar @dudhia
Hugh,
Please take a look a the conflicts that need to be resolved. We are not sure which is what you want. Also, a "stop" statement is not a good idea to use in a physics routine. Use a "call wrf_error_fatal('some error mesage')" construct.

**Comment by davegill** [comment]
@hmorrison100 
Hugh,
Is there a DOI for the paper yet?

**Comment by davegill** [comment]
@dudhia @weiwangncar @hmorrison100 
Folks,
I have a commit that removes the "stop" statements from this PR.
Do we want them? The code would be non-portable.

**Comment by hmorrison100** [comment]
> @hmorrison100
> Hugh,
> Is there a DOI for the paper yet?

Dave, yes the paper is now published. Here it is:
https://doi.org/10.1175/JAS-D-20-0084.1

**Comment by davegill** [comment]
@hmorrison100 
Hugh,
1. There are conflicts in the code that need to be addressed.
2. Would you update the PR message to include the publication citation. I've added the DOI, but it is says "in press".

**Comment by hmorrison100** [comment]
Dave, I'm assuming by "code conflicts" you mean the 2 places that show up in module_mp_p3.F between the master and develop when I click on the button that says "resolve conflicts". In this case, the master version is the correct one.

For the stop statements, did you remove these or replace with something else? The lack of portability is a problem if we replace these with call wrf_error_fatal. We can around this by putting those in a separate if-block to tag as WRF with a "stop" statement otherwise. This is a little clunky but workable. I'm hesitant to just remove the stops without replacing them since the code will crash somewhere else.

**Comment by hmorrison100** [comment]
Also, I just added the full reference of the paper to the PR message.

**Comment by davegill** [comment]
@hmorrison100 
> I'm assuming by "code conflicts" you mean the 2 places that show up in module_mp_p3.F between the master and develop when I click on the button that says "resolve conflicts". In this case, the master version is the correct one.

Thanks, I'll fix it.

> I'm hesitant to just remove the stops 

I do have a replacement for them, but it is not portable. I am inclined to leave this alone right now.

> I just added the full reference of the paper to the PR message.

Thanks for this.

**Comment by hmorrison100** [comment]
Dave, thanks. If you're ok with it I agree it's better to just leave the stops in for now. We can design something more portable for a later update.

**Comment by davegill** [comment]
@hmorrison100 @dudhia @weiwangncar 
Folks,
This PR is ready to review.

**Comment by davegill** [comment]
@kkeene44 
Kelly,
There are a number of updated run_me_first scripts in the P3 PR.

**Comment by davegill** [comment]
@weiwangncar @dudhia 

> Unlike real data case directory, there is only limited physics files that linked to the idealized case directories, and they often depend on the default options used. If we are not linking all of the physics files, we should not add any particular ones, such as those for p3.

Opinions on this?

**Comment by weiwangncar** [comment]
@davegill I did not realize Hugh requested to add these files to those idealized case directories. But my question is still if we link p3* files, should link other physics files?

**Comment by davegill** [comment]
@weiwangncar @dudhia 
> my question is still if we link p3* files, should link other physics files?

We would probably need the scheme data for MP (Thompson, any others?), the radiation schemes (RRTM and RRTMG are used), and the LSM options.

**Comment by weiwangncar** [comment]
@davegill But they are not typically used by the idealized cases... And for those who dare to work on idealized cases, I'd like to assume they know where to get the physics files if needed.

**Comment by hmorrison100** [comment]
@davegill @weiwangncar @dudhia What about the idea of linking files to run the MP schemes for the idealized cases, but not radiation or land surface? This would hopefully keep it manageable, while very few people would run these cases with radiation or land surface turned on.

**Comment by dudhia** [comment]
@hmorrison100 we are still figuring what we want to do in general with ideal cases and linking physics files which is a separate PR.

**Review by dudhia — APPROVED** [review]
I will approve from the physics coupling perspective. Details of file handling not considered for this.

**Pull request merged**